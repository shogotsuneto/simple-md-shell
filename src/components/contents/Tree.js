import React from 'react'
import Section from './Section'
import Heading from './Heading'
import Table from './Table'

const components = {
  root: ({childNodes}) => <React.Fragment>{renderChildNodes(childNodes)}</React.Fragment>,
  section: ({depth, childNodes}) => <Section depth={depth}>{renderChildNodes(childNodes)}</Section>,
  heading: ({depth, childNodes}) => <Heading depth={depth}>{renderChildNodes(childNodes)}</Heading>,
  text: ({value}) => `${value}`,
  paragraph: ({childNodes}) => <p>{renderChildNodes(childNodes)}</p>,
  ruby: ({base, text}) => <ruby>{base}<rt>{text}</rt></ruby>,
  break: () => <br />,
  // table: ({align, childNodes}) => <Table align={align} childNodes={childNodes} renderChildNodes={Tree} />
}

const renderChildNodes = childNodes => childNodes.map((node, index) => <Tree key={`${index}`} node={node} />)

const Tree = ({node}) => {
  const { children: childNodes, ...nodeAttributes} = node
  return components[node.type] ? React.createElement(components[node.type], {...nodeAttributes, childNodes}) : null
}

export default Tree