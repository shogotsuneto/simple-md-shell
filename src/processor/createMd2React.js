import unified from 'unified'
import parse from 'remark-parse'
import sectionize from 'remark-sectionize'
import remarkBreaks from 'remark-breaks'
import remarkAozoraRuby from 'remark-aozora-ruby'
import remark2rehype from 'remark-rehype'
import sanitize from 'rehype-sanitize'
import rehype2react from 'rehype-react'
import gh from 'hast-util-sanitize/lib/github'

const schema = { ...gh, tagNames: [...gh.tagNames, 'section']}

export default ({ createElement, components }) => (
  unified()
    .use(parse)
    .use(sectionize)
    .use(remarkAozoraRuby)
    .use(remarkBreaks)
    .use(remark2rehype)
    .use(sanitize, schema)
    .use(rehype2react, { createElement, components })
    .freeze()
)