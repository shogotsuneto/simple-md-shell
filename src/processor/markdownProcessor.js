import unified from 'unified'
import parse from 'remark-parse'
import sectionize from 'remark-sectionize'
import remarkBreaks from 'remark-breaks'
import remarkAozoraRuby from 'remark-aozora-ruby'

export default unified()
  .use(parse)
  .use(remarkAozoraRuby)
  .use(remarkBreaks)
  .use(sectionize)
  .freeze()