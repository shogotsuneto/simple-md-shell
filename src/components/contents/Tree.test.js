import React from 'react'
import renderer from 'react-test-renderer'

import Tree from './Tree'
import processor from '../../processor/markdownProcessor'

const headers = `
# Header1
## Header2
### Header3
#### Header4
##### Header5
###### Header6
`

const paragraphs = `
paragraph 1

paragraph 2
`

const breaks = `
paragraph  
should break  
should break
`

const ruby = `
振仮名《ふりがな》のテスト。
`

const table = `
| foo | bar | baz |
| :-- | :-: |  -   |
| qux | ｜quux《nested》 | corge |
`

const codeblock = [
  '```js',
  'const md2React = mdText =>',
  '  processor',
  '    .run(processor.parse(mdText))',
  '    .then(node => renderer.create(<Tree node={node} />).toJSON())',
  '```'
].join('\n')

const md2React = mdText =>
  processor
    .run(processor.parse(mdText))
    .then(node => renderer.create(<Tree node={node} />).toJSON())

test('it renders headers', () => {
  md2React(headers).then(tree => expect(tree).toMatchSnapshot())
})

test('it renders paragraphs', () => {
  md2React(paragraphs).then(tree => expect(tree).toMatchSnapshot())
})

test('it renders breaks', () => {
  md2React(breaks).then(tree => expect(tree).toMatchSnapshot())
})

test('it renders ruby', () => {
  md2React(ruby).then(tree => expect(tree).toMatchSnapshot())
})

test('it renders table', () => {
  md2React(table).then(tree => expect(tree).toMatchSnapshot())
})

test('it renders codeblock', () => {
  md2React(codeblock).then(tree => expect(tree).toMatchSnapshot())
})
