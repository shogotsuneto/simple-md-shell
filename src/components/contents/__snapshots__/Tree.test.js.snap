// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders breaks 1`] = `
<p>
  paragraph
  <br />
  should break
  <br />
  should break
</p>
`;

exports[`it renders headers 1`] = `
<section
  className="section1"
>
  <h1
    className="heading1"
  >
    Header1
  </h1>
  <section
    className="section2"
  >
    <h2
      className="heading2"
    >
      Header2
    </h2>
    <section
      className="section3"
    >
      <h3
        className="heading3"
      >
        Header3
      </h3>
      <section
        className="section4"
      >
        <h4
          className="heading4"
        >
          Header4
        </h4>
        <section
          className="section5"
        >
          <h5
            className="heading5"
          >
            Header5
          </h5>
          <section
            className="section6"
          >
            <h6
              className="heading6"
            >
              Header6
            </h6>
          </section>
        </section>
      </section>
    </section>
  </section>
</section>
`;

exports[`it renders paragraphs 1`] = `
Array [
  <p>
    paragraph 1
  </p>,
  <p>
    paragraph 2
  </p>,
]
`;

exports[`it renders ruby 1`] = `
<p>
  <ruby>
    振仮名
    <rt>
      ふりがな
    </rt>
  </ruby>
  のテスト。
</p>
`;

exports[`it renders table 1`] = `
<table>
  <thead>
    <tr>
      <th
        className="cell-align-left"
      >
        foo
      </th>
      <th
        className="cell-align-center"
      >
        bar
      </th>
      <th
        className="cell-align-null"
      >
        baz
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="cell-align-left"
      >
        qux
      </td>
      <td
        className="cell-align-center"
      >
        quux
      </td>
      <td
        className="cell-align-null"
      >
        corge
      </td>
    </tr>
  </tbody>
</table>
`;
